### PR TITLE
Don't listen for both "click" and "pointerup" event.

### PR DESCRIPTION
**Pull request by breville**
For a long time, we've been executing the same handler for both the "click" and "touchend/MSPointerUp/pointerup" event for a user action.  At least on IE11, the use of preventDefault for the latter event did not prevent the generation of a click event as well.  (Nor, for that matter, did an attempt at stopPropagation.  I think preventDefault would have worked on touchstart, but we probably chose touchend for a reason.)  

It seems that the effect of this was masked until now, until https://github.com/code-dot-org/code-dot-org/pull/2336 added the cool feature to re-open instructions.  On IE11, this was opening two copies of the instructions and some remnants of the second set were left behind when the instructions were re-dismissed.

The fix is to not listen for the "click" event if there is a touch/pointer event that we handle instead.

Tested on desktop Chrome, iOS Safari, and on BrowserStack for IE10 & 11 desktop & metro.

### CODE REVIEW

**Comment by breville** [comment]
Scary old click/touch code... any thoughts from @Bjvanminnen @davidsbailey @bcjordan @joshlory @cpirich ?  Are there additional ways to test this?

**Comment by Bjvanminnen** [comment]
We have a bunch of physical devices that are probably worth testing real touch on.

File: apps/src/dom.js [hunk 1]
```diff
@@ -37,14 +37,14 @@ exports.getTouchEventName = function(eventName) {
 };
 
 var addEvent = function(element, eventName, handler) {
-  element.addEventListener(eventName, handler, false);
-
   var touchEvent = exports.getTouchEventName(eventName);
   if (touchEvent) {
```
Comment (joshlory) [inline apps/src/dom.js]: The challenge is that touch and mouse aren't mutually exclusive.  E.g. the Chromebook Pixel supports touch events on its touchscreen but also has a trackpad.  If we don't also bind mouse events we'll break mouse interaction for these devices.

Comment (bcjordan) [inline apps/src/dom.js]: Would replacing the whole handler with a technique like this maybe work? http://stackoverflow.com/a/11507558/136134

Comment (breville) [inline apps/src/dom.js]: Yeah, my next attempt is just changing it to hook touchstart/MSPointerDown/pointerdown in which case preventDefault should eat the event the first time.  Downside is that drag events out of that UI element will fire as a fast click, but on the plus side it means we now have fast click, and this should be localised to the elements that are bound this way.

Comment (cpirich) [inline apps/src/dom.js]: yeah, I'd be careful here. I just modified this code a little recently so I could enable dragging of the resize bars (to change visualization column size for all apps or debug area size for applab only). Having it register both mousedown and touchdown may be important for my use cases.

One question which I've never fully understood is why do some of these buttons need this code path at all? Can't you just listen for "click"?

Comment (davidsbailey) [inline apps/src/dom.js]: click has a 300ms delay on touch devices, so we listen for touchstart to have a snappier experience.

Comment (davidsbailey) [inline apps/src/dom.js]: generally this looks like a very broad-reaching change and makes me nervous. what is the user-facing problem we are trying to fix here? or is this a tech debt cleanup issue?

Comment (joshlory) [inline apps/src/dom.js]: There are a few options for opting out of the 300ms click delay here: http://developer.telerik.com/featured/300-ms-click-delay-ios-8/

**Comment by breville** [comment]
I've updated with a second attempt.  Comments from that commit:

In this approach, we don't get a double call to our handler, but we do it by first looking for a touchstart.  If we get one, then we eat the event (which wasn't possibly when looking for touchend), otherwise the regular click handler will kick in.

The main downside is that the user doesn't get to touch a clickable item and drag the screen around before the button press occurs when the touch ends.  In this new approach, as soon as the clickable item is touched, the action is taken.  This is arguably more correct, and certainly more in the "fastclick" spirit.

I haven't tried with every single button, but a search for dom.addClickTouchEvent in apps/src looks like everything that gets this treatment is some kind of button that should benefit from the "fastclick" approach.

**Comment by davidsbailey** [comment]
I just realized that 'click' already gets fired on touch devices, so the only point in using addClickTouchEvent is to get a faster click response. Therefore, a safe fallback option would just be to bind to the click handler without binding any special MS or Touch event. Is that your understanding as well Brendan? 

LGTM. However, it would be helpful if you could paste a quick list of features you tested which this affects, so that anyone investigating a test failure can link it back to this PR.

**Comment by breville** [comment]
It seems at least bounce explicitly uses dom.addMouseDownTouchEvent for button down and dom.addClickTouchEvent for button up.  I'm scared other apps might do similar, so I think I will just make a surgical change to the instructions dialog display in studioApp.js.

**Comment by davidsbailey** [comment]
sounds reasonable. Can you also add a comment to addClickTouchEvent like: "DEPRECATED: this method can cause events to fire multiple times on touch devices."

**Comment by breville** [comment]
Actually I decided to go ahead and change dom.addClickTouchEvent to use the beginning of the mouse/touch event rather than the end.  This gives us "fastclick" style behaviour for many more buttons.

I also modified bounce's setup of its arrow buttons to explicitly use dom.addMouseUpTouchEvent instead of dom.addClickTouchEvent so that it catches the button/touch up event explicitly.

Testing: I tested the instructions dismissing/restoring/dismissing, and the left/right arrow buttons on a bounce level (/s/course3/stage/15/puzzle/7), on OSX Chrome, iOS Safari, a real Android device, a real Surface, and in IE10 desktop & IE 11 desktop on Browserstack.

**Pull request merged**